app/models/elastic_search.rb

```diff
@@ -0,0 +1,314 @@
+    # Eg: worng spelling: 'hiden figurs', Still fuzziness will give results like "Hidden Figures"
+    if !results[:hits].present? && params["keyword"].present? && query[:query][:bool][:must].present?
+      if query[:query][:bool][:must][0][:multi_match].present?
+        query[:query][:bool][:must][0][:multi_match][:fuzziness] = 1
+        results = search_by_query(query)
+      end
+    end
+    results
+  end
+
+
+  # Get elastic serach queries based on input filter params.
+  def teacher_sets_query_based_on_filters(params)
+    keyword, grade_begin, grade_end, language, set_type, availability, area_of_study, subjects = teacher_sets_input_params(params)
+    query = {:query => {:bool => {:must => []}}}
+
+    # If search keyword is present in filters, finding the search keyword in these fields [title, description, contents]
+    if keyword.present?
+      query[:query][:bool][:must] << {:multi_match => {:query => keyword, :fields => %w[title^8 description contents]}}
+    end
+
+    # If grade_begin, grade_end ranges present in filters get ES query based on ranges.
+    # grade_begin value should be less than grade_end value
+    # grade_end value should be greater than grade_begin value
+    if grade_begin.present? && grade_end.present?
+      query[:query][:bool][:must] << {:range => {:grade_begin => {:lte => grade_end.to_i}}}
+      query[:query][:bool][:must] << {:range => {:grade_end => {:gte => grade_begin.to_i}}}
+    end
+
+    # If language present in filters finding the language in these fields [language, primary_language]
+    if language.present?
+      query[:query][:bool][:must] << {:multi_match => {:query => language.join, :fields => %w[language primary_language]}}
+    end
+
+    # If set_type present in filters get ES query based on set_type.
+    # Eg: set_type: single/multi
+
+    if set_type.present?
+      query[:query][:bool][:must] << {:match => {:set_type => set_type.join}}
+    end
+
+    # If availability present in filters get ES query based on availability.
+    # Eg: availability: "available/unavailable"
+    if availability.present?
+      query[:query][:bool][:must] << {:match => {:availability => availability.join}}
+    end
+
+    # If area_of_study present in filters get ES query based on area_of_study.
+    # Eg: area_of_study: "Social Studies"
+    if area_of_study.present?
+      query[:query][:bool][:must] << {:match => {:area_of_study => area_of_study.join}}
+    end
+
+    # If subjects present in filters get ES query based on subjects.
+    # teacherset have has_many  relationship with subject.
+    # subjects mapping are stored in nested format in elastic search. 
+    if subjects.present?
+      query[:query][:bool][:must] << {:nested => {:path => "subjects", :query => {:bool => {:must => {:terms => {"subjects.id" => subjects}}}}}}
+    end
+    query
+  end
+
+  
+  # Get teacher set facets
+  def facets_for_query(_teacher_sets)
+    # TODO: Need to work on tha rails cache
```

Comment:
is that something that needs to be done before this code is published, or should we add a backlog item?

Reply:
@daryanypl Yes i am working on it.